### DISCUSSION TITLE
Group license raises `RuntimeError: request license token faild (3)`

### DISCUSSION THREAD

**Discussion by alikaz3mi** [dashingsoft/pyarmor#1425] [Q&A] [answered]
I have upgraded pyarmor to 8.3.3. But the issue still exists:

```
(py39) deepdev@deepdev-H270-Gaming-3:~$ pyarmor -v
Pyarmor 8.3.3 (group), 006070, dashboard

License Type    : pyarmor-group
License No.     : pyarmor-vax-006070
License To      : HEMSCap
License Product : dashboard

BCC Mode        : Yes
RFT Mode        : Yes

Notes
* Offline obfuscation

(py39) deepdev@deepdev-H270-Gaming-3:~$ unzip pyarmor-device-regfile-6070.5.zip -d xx
Archive:  pyarmor-device-regfile-6070.5.zip
replace xx/license.lic? [y]es, [n]o, [A]ll, [N]one, [r]ename: A
 extracting: xx/license.lic          
 extracting: xx/.pyarmor_capsule.zip  
 extracting: xx/group.tokens         
 extracting: xx/tokens/g04831288b9fb495661ffffd1eb2c0c52  
(py39) deepdev@deepdev-H270-Gaming-3:~$ python
Python 3.9.13 (main, Aug 25 2022, 23:26:10) 
[GCC 11.2.0] :: Anaconda, Inc. on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from pyarmor.cli.core import Pytransform3
>>> print('machine id is "%s"' % Pytransform3.get_hd_info(16))
machine id is "b'g04831288b9fb495661ffffd1eb2c0c52'"
```

Error during running `pyarmor gen -O encoded -r my_project

```
(py39) deepdev@deepdev-H270-Gaming-3:~/PycharmProjects/FaceRecognition$ cat pyarmor.error.log 
375921 MainProcess 2023-08-15 12:50:17,732
Traceback (most recent call last):
  File "/home/deepdev/anaconda3/envs/py39/lib/python3.9/site-packages/pyarmor/cli/__main__.py", line 702, in main
  File "/home/deepdev/anaconda3/envs/py39/lib/python3.9/site-packages/pyarmor/cli/__main__.py", line 690, in main_entry
  File "/home/deepdev/anaconda3/envs/py39/lib/python3.9/site-packages/pyarmor/cli/__main__.py", line 229, in cmd_gen
  File "/home/deepdev/anaconda3/envs/py39/lib/python3.9/site-packages/pyarmor/cli/generate.py", line 156, in process
  File "/home/deepdev/anaconda3/envs/py39/lib/python3.9/site-packages/pyarmor/cli/generate.py", line 119, in _obfuscate_scripts
  File "/home/deepdev/anaconda3/envs/py39/lib/python3.9/site-packages/pyarmor/cli/core/__init__.py", line 103, in generate_obfuscated_script
  File "<maker>", line 665, in generate_obfuscated_script
  File "/home/deepdev/anaconda3/envs/py39/lib/python3.9/site-packages/pyarmor/cli/__init__.py", line 16, in process
  File "<maker>", line 504, in process
  File "<maker>", line 510, in coserialize
  File "<maker>", line 555, in _build_ast_body
RuntimeError: request license token faild (3)
```

**Comment by jondy** [comment]
What's the output log of this command?
```
pyarmor -d reg pyarmor-device-regfile-6070.5.zip
```

**Comment by jondy** [comment]
And what's the full output log to obfuscate simple script with debug option?
```
echo "print('hello')" > foo.py
pyarmor -d gen --enable-rft foo.py
```

**Comment by alikaz3mi** [comment]
```
pyarmor -d reg pyarmor-device-regfile-6070.5.zip 
INFO     Python 3.9.13
INFO     Pyarmor 8.3.3 (group), 006070, dashboard
INFO     Platform linux.x86_64
DEBUG    native platform linux.x86_64
DEBUG    home path: /home/deepdev/.pyarmor
INFO     register "pyarmor-device-regfile-6070.5.zip"
DEBUG    extracting license.lic
DEBUG    extracting .pyarmor_capsule.zip
DEBUG    group license for machines: ['tokens/g04831288b9fb495661ffffd1eb2c0c52']
DEBUG    got machine id: g94c8eca511e061d8c49bfe2dbbd953b4
DEBUG    got machine id: bca27cc53bebd099f065e9b2d3d7c178f
ERROR    could not get docker host machine id
Traceback (most recent call last):
  File "/home/deepdev/anaconda3/envs/py39/lib/python3.9/site-packages/pyarmor/cli/register.py", line 194, in _get_docker_hostname
    s.connect((host, port))
socket.gaierror: [Errno -2] Name or service not known
ERROR    could not get docker host machine id
```


I have just used device id file 5. Now, I can't use it!

```
pyarmor -d gen --enable-rft foo.py
INFO     Python 3.9.13
INFO     Pyarmor 8.3.3 (group), 006070, dashboard
INFO     Platform linux.x86_64
DEBUG    native platform linux.x86_64
DEBUG    home path: /home/deepdev/.pyarmor
DEBUG    command options: {'no_runtime': False, 'enable_rft': True, 'inputs': ['foo.py']}
DEBUG    install plugin: CodesignPlugin
DEBUG    install plugin: DarwinUniversalPlugin
INFO     search inputs ...
INFO     find script foo.py
INFO     find 1 top resources
DEBUG    unknown error, please check pyarmor.error.log
ERROR    request license token faild (3)
```

I haven't changed my machine

**Reply by alikaz3mi** [reply]
At first, I thought that it is related to different anaconda environments and system environment. But testing it on another machine lead to the same result: could not get docker host machine id

**Comment by jondy** [comment]
> DEBUG    group license for machines: ['tokens/g04831288b9fb495661ffffd1eb2c0c52']
DEBUG    got machine id: g94c8eca511e061d8c49bfe2dbbd953b4
DEBUG    got machine id: bca27cc53bebd099f065e9b2d3d7c178f

The machine id is different.

Try to generate device file in each anaconda environments.

**Comment by jondy** [comment]
Make sure computer name doesn't change in different anaconda environments
Check computer name by Python in each anaconda environment
```
python -c"import os
print(os.uname().nodename)
```

**Comment by alikaz3mi** [comment]
It is the same on all of my conda environments and system's python.

I've update pyarmor from 8.3.1 to 8.3.3 on my other device which is a virtual machine. But I am still getting the same error. (license token failed(3).

But, at least pyarmor -d gen --enable-rft foo.py seems to be working.

```
pyarmor -d gen --enable-rft foo.py
INFO     Python 3.8.10
INFO     Pyarmor 8.3.3 (group), 006070, dashboard
INFO     Platform linux.x86_64
DEBUG    native platform linux.x86_64
DEBUG    home path: /home/dashvm/.pyarmor
DEBUG    command options: {'no_runtime': False, 'enable_rft': True, 'inputs': ['foo.py']}
DEBUG    install plugin: CodesignPlugin
DEBUG    install plugin: DarwinUniversalPlugin
INFO     search inputs ...
INFO     find script foo.py
INFO     find 1 top resources
DEBUG    build package relations
INFO     start to generate runtime files
INFO     target platforms {'linux.x86_64'}
INFO     write dist/pyarmor_runtime_006070/pyarmor_runtime.so
INFO     patch runtime file
DEBUG    patching runtime data at 764096
DEBUG    patch runtime file OK
INFO     call post runtime plugin <class 'pyarmor.cli.plugin.CodesignPlugin'>
INFO     generate runtime files OK
INFO     start to obfuscate scripts
INFO     process resource "foo"
INFO     obfuscating file foo.py
DEBUG    process inline marker
DEBUG    parse script
DEBUG    process co
DEBUG    patch co
INFO     write dist/foo.py
INFO     obfuscate scripts OK
INFO     call post build plugin <class 'pyarmor.cli.plugin.DarwinUniversalPlugin'>
```

**Comment by jondy** [comment, marked as answer]
Could you check machine id for each conda environments
```
(py39) deepdev@deepdev-H270-Gaming-3:~$ python
Python 3.9.13 (main, Aug 25 2022, 23:26:10) 
[GCC 11.2.0] :: Anaconda, Inc. on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from pyarmor.cli.core import Pytransform3
>>> print('machine id is "%s"' % Pytransform3.get_hd_info(16))
machine id is "b'g04831288b9fb495661ffffd1eb2c0c52'"
```
 
Here the solution for group license raises “ERROR request license token failed”

First upgrade Pyarmor to 8.3.3+

Then register group license with debug option -d in offline device. For example:
```
$ pyarmor -d reg pyarmor-device-regfile-6000.4.zip
```
Check log, make sure current machine id is inclueded by offline regfile. For example:
```
DEBUG    group license for machines: ['tokens/gb04eb35da4f5378185c8663522e0a5e3']
DEBUG    got machine id: gb04eb35da4f5378185c8663522e0a5e3
```
If machine id is mismatched, please generate new device file for this device by Pyarmor 8.3.3+

For virtual machine, make sure machine id is same after reboot.

**Reply by alikaz3mi** [reply]
I updated the pyarmor.
The machine id on all of the conda environments and the system's python are the same.


The machine id changed when I restarted my system and the virtual machine.

**Comment by jondy** [comment]
What's your system? It's physical device or something else.

If you're using group license in virtual machine or any system, make sure computer name, Mac address of network card, and hard disk (virtual) doesn't change. If they're changed, group license is not designed to work for this case.

If only need run the obfuscated scripts in virtual machines, try to obfuscate the scripts in physical device or any system which hardware are same, then copy the obfuscated scripts to virtual machines.

**Reply by alikaz3mi** [reply]
It is a physical Desktop PC device. I haven't changed my system since I have installed pyarmor. I am using group license. I wanted to use the virtual machine for obfuscation. It is a physical Desktop PC device. I haven't changed my system since I have installed pyarmor. I am using group license. I wanted to use the virtual machine for obfuscation.

**Reply by alikaz3mi** [reply]
I updated to pyarmor 8.3.4. The issue still exists.

**Reply by alikaz3mi** [reply]
deepdev@deepdev-H270-Gaming-3:~$ pyarmor -d reg  pyarmor-device-regfile-6070.7.zip 
INFO     Python 3.10.12
INFO     Pyarmor 8.3.4 (group), 006070, dashboard
INFO     Platform linux.x86_64
DEBUG    native platform linux.x86_64
DEBUG    home path: /home/deepdev/.pyarmor
INFO     register "pyarmor-device-regfile-6070.7.zip"
DEBUG    extracting license.lic
DEBUG    extracting .pyarmor_capsule.zip
DEBUG    group license for machines: ['tokens/g317cdf11dfd1ebadac2f8bd58aee6e20']
DEBUG    got machine id: g317cdf11dfd1ebadac2f8bd58aee6e20
DEBUG    extracting tokens/g317cdf11dfd1ebadac2f8bd58aee6e20
INFO     This license registration information:

License Type    : pyarmor-group
License No.     : pyarmor-vax-006070
License To      : HEMSCap
License Product : dashboard

BCC Mode        : Yes
RFT Mode        : Yes

Notes
* Offline obfuscation

**Comment by jondy** [comment]
First of all, please run all the pyarmor commands by same python interpreter.
If obfuscating the scripts in anaconda environments, please run `pyarmor -d reg xxx.zip` also in anaconda environments.

I have 2 questions,
1. Is machine id of physical Desktop PC device same after reboot?
2. Is machine id got by system python same as got in anaconda environments?

**Comment by jondy** [comment]


**Reply by alikaz3mi** [reply]
On the other Desktop pc and my I own, I have ran the command using the same python interpreter. 

1. In all of the 3 devices (2 Desktop PC, 1 virtual environment), the machine id from  `Pytransform3.get_hd_info(16)` changed after restartng the machine. 
2. Yes, the machine id is equal between the system's python and conda environments

**Comment by alikaz3mi** [comment]
I haven't restarted my system for several days. All of a sudden pyarmor -v shows that my license is trial. And the machine id (`Pytransform3.get_hd_info(16)) is different from what it was the other days!

**Comment by jondy** [comment]
I may build debug version to trace machine id, please following the progress in the issue #1446, the download link will be shared in issue #1446.

**Reply by alikaz3mi** [reply]
Sure thing
